File: Assets/Bundles/Shared/Resources/TS/Entity/Animation/InventoryEntityAnimator.lua
-- Compiled with unity-ts v2.1.0-75
local BundleReferenceManager = require("Shared/TS/Util/BundleReferenceManager").BundleReferenceManager
local MathUtil = require("Shared/TS/Util/MathUtil").MathUtil
local BundleGroupNames = require("Shared/TS/Util/ReferenceManagerResources").BundleGroupNames
local EntityAnimator = require("Shared/TS/Entity/Animation/EntityAnimator").EntityAnimator
local ItemEventKeys
do
	local _inverse = {}
	ItemEventKeys = setmetatable({}, {
		__index = _inverse,
	})
	ItemEventKeys.NONE = -1
	_inverse[-1] = "NONE"
	ItemEventKeys.IDLE = 0
	_inverse[0] = "IDLE"
	ItemEventKeys.EQUIP = 1
	_inverse[1] = "EQUIP"
	ItemEventKeys.UN_EQUIP = 2
	_inverse[2] = "UN_EQUIP"
	ItemEventKeys.USE = 3
	_inverse[3] = "USE"
end
local ItemPlayMode
do
	local _inverse = {}
	ItemPlayMode = setmetatable({}, {
		__index = _inverse,
	})
	ItemPlayMode.DEFAULT = 0
	_inverse[0] = "DEFAULT"
	ItemPlayMode.LOOP = 1
	_inverse[1] = "LOOP"
	ItemPlayMode.HOLD = 2
	_inverse[2] = "HOLD"
end
local InventoryEntityAnimator
do
	local super = EntityAnimator
	InventoryEntityAnimator = setmetatable({}, {
		__tostring = function()
			return "InventoryEntityAnimator"
		end,
		__index = super,
	})
	InventoryEntityAnimator.__index = InventoryEntityAnimator
	function InventoryEntityAnimator.new(...)
		local self = setmetatable({}, InventoryEntityAnimator)
		return self:constructor(...) or self
	end
	function InventoryEntityAnimator:constructor(entity, anim, ref)
		super.constructor(self, entity, anim, ref)
		self.itemLayerIndex = 2
		self.spineClampAngle = 15
		self.neckClampAngle = 35
		self.currentItemClips = {}
		self.bundleIndex = 0
		self.currentBundleName = BundleGroupNames.NONE
		self.currentItemState = ItemEventKeys.NONE
		self.isFirstPerson = false
		self.itemLayer = AnimancerBridge:GetLayer(self.anim, self.itemLayerIndex)
		-- Initial animation setup
		self:LoadNewItemResources(BundleGroupNames.ItemUnarmed)
		self:SetFirstPerson(false)
	end
	function InventoryEntityAnimator:LateUpdate()
		if not self.isFirstPerson then
			self:ForceLookForward()
		end
	end
	function InventoryEntityAnimator:Log(message)
		return nil
	end
	function InventoryEntityAnimator:ForceLookForward()
		self:ClampRotation(self.entityRef.spineBone1, self.spineClampAngle)
		self:ClampRotation(self.entityRef.spineBone2, self.spineClampAngle)
		-- this.ClampRotation(this.entityRef.spineBone3, this.spineClampAngle);
		self:ClampRotation(self.entityRef.neckBone, self.neckClampAngle)
	end
	function InventoryEntityAnimator:ClampRotation(spine, maxAngle)
		-- Take the world look and convert to this spines local space
		local _exp = Quaternion.Inverse(spine.parent.rotation)
		local _rotation = self.entityRef.root.rotation
		local targetLocalRot = _exp * _rotation
		-- Clamp the rotation so the spine doesn't appear broken
		local rotY = MathUtil:ClampAngle(targetLocalRot.eulerAngles.y, -maxAngle, maxAngle)
		spine.localEulerAngles = Vector3.new(spine.localEulerAngles.x, rotY, spine.localEulerAngles.z)
	end
	function InventoryEntityAnimator:SetFirstPerson(isFirstPerson)
		self.isFirstPerson = isFirstPerson
		self.bundleIndex = if isFirstPerson then 0 else 1
		if self.currentBundleName ~= BundleGroupNames.NONE then
			-- First person and third person use different animation bundles
			-- So we need to load the item resources again
			self:LoadNewItemResources(self.currentBundleName)
			self:StartItemIdle()
		end
	end
	function InventoryEntityAnimator:Play(clipKey, onEnd, wrapMode)
		if wrapMode == nil then
			wrapMode = 0
		end
		local clip = self.currentItemClips[clipKey + 1]
		if clip == nil then
			-- No animation for this event
			self.itemLayer:StartFade(0, self.defaultTransitionTime)
			return nil
		end
		self:Log("Playing Item Anim: " .. tostring(clipKey))
		self.itemLayer:StartFade(1, self.defaultTransitionTime)
		local state = self:PlayAnimation(clip, self.itemLayerIndex, wrapMode)
		if onEnd ~= nil then
			state.Events:OnEndTS(onEnd)
		end
	end
	function InventoryEntityAnimator:LoadNewItemResources(nextItemId)
		self:Log("Loading Item: " .. tostring(nextItemId))
		self.currentBundleName = nextItemId
		self.itemLayer:DestroyStates()
		-- Load the animation clips for the new item
		self.currentItemClips = BundleReferenceManager:LoadResources(nextItemId, self.bundleIndex)
	end
	function InventoryEntityAnimator:TriggerEvent(key, index)
		if index == nil then
			index = 0
		end
		self:Log("Trigger State: " .. tostring(key) .. " index: " .. tostring(index))
		self.currentItemState = key
		--[[
			eventData.eventKey = key;
			eventData.eventIndex = index;
			currentItem.OnAnimEvent(eventData);
		]]
	end
	function InventoryEntityAnimator:EquipItem(itemId)
		-- Have to animate out the current item before the new item can be added
		self:StartUnEquipAnim(itemId)
	end
	function InventoryEntityAnimator:StartUnEquipAnim(nextItemId)
		-- Right now we are ignoring unequipped and loading the new item instantly
		self:LoadNewItemResources(nextItemId)
		self:StartItemEquipAnim()
		--[[
			//Play the un-equip animation
			this.TriggerEvent(ItemEventKeys.UN_EQUIP);
			this.Play(ItemEventKeys.UN_EQUIP, () => {
			//Load the resources for the next item
			//this.LoadNewItemResources(nextItemId);
			this.StartItemEquipAnim();
			});
		]]
	end
	function InventoryEntityAnimator:StartItemEquipAnim()
		self:TriggerEvent(ItemEventKeys.EQUIP)
		self:Play(ItemEventKeys.EQUIP, function()
			self:StartItemIdle()
		end)
	end
	function InventoryEntityAnimator:StartItemIdle()
		self:TriggerEvent(ItemEventKeys.IDLE)
		self:Play(ItemEventKeys.IDLE)
	end
	function InventoryEntityAnimator:PlayItemUse(useIndex, itemPlayMode)
		if useIndex == nil then
			useIndex = 0
		end
		if itemPlayMode == nil then
			itemPlayMode = 0
		end
		-- In the animation array use animations are the 3rd index and beyond;
		local i = useIndex + 3
		if i >= 0 and i < #self.currentItemClips then
			self:TriggerEvent(ItemEventKeys.USE, useIndex)
			self:Play(i, function()
				if itemPlayMode == ItemPlayMode.DEFAULT then
					self:StartItemIdle()
				elseif itemPlayMode == ItemPlayMode.LOOP then
					self:PlayItemUse(useIndex, ItemPlayMode.LOOP)
				end
			end)
		else
			warn("Trying to play animation that doesn't exist: use " .. tostring(useIndex))
		end
	end
end
return {
	ItemEventKeys = ItemEventKeys,
	ItemPlayMode = ItemPlayMode,
	InventoryEntityAnimator = InventoryEntityAnimator,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Entity/Animation/InventoryEntityAnimator.ts
import { RunUtil } from "Shared/Util/RunUtil";
import { BundleReferenceManager } from "../../Util/BundleReferenceManager";
import { MathUtil } from "../../Util/MathUtil";
import { BundleGroupNames } from "../../Util/ReferenceManagerResources";
import { Entity, EntityReferences } from "../Entity";
import { EntityAnimator } from "./EntityAnimator";

export enum ItemEventKeys {
	NONE = -1,
	IDLE,
	EQUIP,
	UN_EQUIP,
	USE,
}

export enum ItemPlayMode {
	DEFAULT,
	LOOP,
	HOLD,
}

export class InventoryEntityAnimator extends EntityAnimator {
	private readonly itemLayerIndex: number = 2;
	private itemLayer: AnimancerLayer;
	private spineClampAngle = 15;
	private neckClampAngle = 35;

	private currentItemClips: Array<AnimationClip> = [];
	private bundleIndex = 0;
	private currentBundleName: BundleGroupNames = BundleGroupNames.NONE;
	private currentItemState: ItemEventKeys = ItemEventKeys.NONE;
	private isFirstPerson = false;

	//private camera: Camera;

	public constructor(entity: Entity, anim: AnimancerComponent, ref: EntityReferences) {
		super(entity, anim, ref);
		this.itemLayer = AnimancerBridge.GetLayer(this.anim, this.itemLayerIndex);
		//Initial animation setup
		this.LoadNewItemResources(BundleGroupNames.ItemUnarmed);
		this.SetFirstPerson(false);
	}

	public LateUpdate() {
		if (!this.isFirstPerson && RunUtil.IsClient()) {
			this.ForceLookForward();
		}
	}

	private Log(message: string) {
		return;
		print("Animator " + this.entity.id + ": " + message);
	}

	//Always keep the character looking where the player is looking
	private ForceLookForward() {
		this.ClampRotation(this.entityRef.spineBone1, this.spineClampAngle);
		this.ClampRotation(this.entityRef.spineBone2, this.spineClampAngle);
		//this.ClampRotation(this.entityRef.spineBone3, this.spineClampAngle);
		this.ClampRotation(this.entityRef.neckBone, this.neckClampAngle);
	}

	private ClampRotation(spine: Transform, maxAngle: number) {
		//Take the world look and convert to this spines local space
		let targetLocalRot = Quaternion.Inverse(spine.parent.rotation).mul(this.entityRef.root.rotation);

		//Clamp the rotation so the spine doesn't appear broken
		let rotY = MathUtil.ClampAngle(targetLocalRot.eulerAngles.y, -maxAngle, maxAngle);
		spine.localEulerAngles = new Vector3(spine.localEulerAngles.x, rotY, spine.localEulerAngles.z);
	}

	public SetFirstPerson(isFirstPerson: boolean) {
		this.isFirstPerson = isFirstPerson;
		this.bundleIndex = isFirstPerson ? 0 : 1;
		if (this.currentBundleName !== BundleGroupNames.NONE) {
			//First person and third person use different animation bundles
			//So we need to load the item resources again
			this.LoadNewItemResources(this.currentBundleName);
			this.StartItemIdle();
		}
	}

	private Play(clipKey: ItemEventKeys, onEnd?: Callback, wrapMode: WrapMode = WrapMode.Default) {
		let clip = this.currentItemClips[clipKey];
		if (clip === undefined) {
			//No animation for this event
			this.itemLayer.StartFade(0, this.defaultTransitionTime);
			return;
		}
		this.Log("Playing Item Anim: " + clipKey);
		this.itemLayer.StartFade(1, this.defaultTransitionTime);
		const lastState = this.itemLayer.CurrentState;
		if (lastState) {
			//Clear the last states end event since we are now starting a new animation
			lastState.Events.ClearEndTSEvent();
		}
		const animState = this.PlayAnimation(clip, this.itemLayerIndex, wrapMode);
		if (onEnd !== undefined) {
			animState.Events.OnEndTS(onEnd);
		}
	}

	private LoadNewItemResources(nextItemId: BundleGroupNames) {
		this.Log("Loading Item: " + nextItemId);
		this.currentBundleName = nextItemId;
		this.itemLayer.DestroyStates();
		//Load the animation clips for the new item
		this.currentItemClips = BundleReferenceManager.LoadResources<AnimationClip>(nextItemId, this.bundleIndex);
	}

	//When certain conditions are met, the items may want to play their own synced animations
	//This was used in testing to sync that
	//private HumanoidItemAnimEventData eventData = new ();
	private TriggerEvent(key: ItemEventKeys, index = 0) {
		this.Log("Trigger State: " + key + " index: " + index);
		this.currentItemState = key;
		/*eventData.eventKey = key;
        eventData.eventIndex = index;
        currentItem.OnAnimEvent(eventData);*/
	}

	public EquipItem(itemId: BundleGroupNames) {
		//Have to animate out the current item before the new item can be added
		this.StartUnEquipAnim(itemId);
	}

	private StartUnEquipAnim(nextItemId: BundleGroupNames) {
		//Right now we are ignoring unequipped and loading the new item instantly
		this.LoadNewItemResources(nextItemId);
		this.StartItemEquipAnim();

		/*
		//Play the un-equip animation
		this.TriggerEvent(ItemEventKeys.UN_EQUIP);
		this.Play(ItemEventKeys.UN_EQUIP, () => {
			//Load the resources for the next item
			//this.LoadNewItemResources(nextItemId);
			this.StartItemEquipAnim();
		});
		*/
	}

	private StartItemEquipAnim() {
		this.TriggerEvent(ItemEventKeys.EQUIP);
		this.Play(ItemEventKeys.EQUIP, () => {
			this.StartItemIdle();
		});
	}

	public StartItemIdle() {
		this.TriggerEvent(ItemEventKeys.IDLE);
		this.Play(ItemEventKeys.IDLE);
	}

	public PlayItemUse(useIndex = 0, itemPlayMode: ItemPlayMode = 0) {
		this.Log("Item Use Started: " + useIndex);
		//In the animation array use animations are the 3rd index and beyond;
		let i = useIndex + 3;
		if (i >= 0 && i < this.currentItemClips.size()) {
			this.TriggerEvent(ItemEventKeys.USE, useIndex);
			this.Play(i, () => {
				if (itemPlayMode === ItemPlayMode.DEFAULT) {
					this.StartItemIdle();
				} else if (itemPlayMode === ItemPlayMode.LOOP) {
					this.PlayItemUse(useIndex, ItemPlayMode.LOOP);
				}
			});
		} else {
			warn("Trying to play animation that doesn't exist: use " + useIndex);
		}
	}
}
